src/compiler/utilities.ts

```diff
@@ -3932,32 +3932,83 @@ namespace ts {
 
     export function createTextWriter(newLine: string): EmitTextWriter {
         let output: string;
+        const pendingCharCodes: number[] = [];
+        let lastChar: number;
         let indent: number;
         let lineStart: boolean;
         let lineCount: number;
         let linePos: number;
+        let totalChars: number = 0;
         let hasTrailingComment = false;
 
-        function updateLineCountAndPosFor(s: string) {
-            const lineStartsOfS = computeLineStarts(s);
-            if (lineStartsOfS.length > 1) {
-                lineCount = lineCount + lineStartsOfS.length - 1;
-                linePos = output.length - s.length + last(lineStartsOfS);
-                lineStart = (linePos - output.length) === 0;
+        const newLineCodes: number[] = [];
+        for (let i = 0; i < newLine.length; i++) {
+            newLineCodes.push(newLine.charCodeAt(i));
+        }
+
+        function appendCharCode(charCode: number) {
+            ++totalChars;
+            lastChar = charCode;
+            pendingCharCodes.push(charCode);
+            if (pendingCharCodes.length >= 1024) {
+                flushBuffer();
             }
-            else {
-                lineStart = false;
+        }
+
+        function flushBuffer() {
+            if (pendingCharCodes.length > 0) {
+                output += String.fromCharCode.apply(null, pendingCharCodes);
+                pendingCharCodes.length = 0;
             }
         }
 
+        function appendRawSmall(text: string) {
+            const len = text.length;
+            for (let pos = 0; pos < len; pos++) {
+                const ch = text.charCodeAt(pos);
+                appendCharCode(ch);
+                // Ignore carriageReturn, since we mark the following lineFeed as the newline anyway
+                if (isLineBreak(ch) && ch !== CharacterCodes.carriageReturn) {
+                    ++lineCount;
+                    linePos = totalChars;
+                }
+            }
+
+            lineStart = linePos === totalChars;
+        }
+
+        function appendRawLarge(text: string) {
+            flushBuffer();
+
+            const len = text.length;
+            for (let pos = 0; pos < len; pos++) {
+                const ch = text.charCodeAt(pos);
+                ++totalChars;
+                lastChar = ch;
+                // Ignore carriageReturn, since we mark the following lineFeed as the newline anyway
+                if (isLineBreak(ch) && ch !== CharacterCodes.carriageReturn) {

```

Comment:
Technically carriage return can appear on its own and will be considered a newline.